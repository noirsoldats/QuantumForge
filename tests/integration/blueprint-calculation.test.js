/**
 * Integration Tests for Blueprint Calculation Flow
 *
 * Tests complete end-to-end blueprint calculation workflows
 * using in-memory database and realistic data
 */

const { calculateBlueprintMaterials } = require('../../src/main/blueprint-calculator');
const { createInMemoryDatabase, populateDatabase } = require('../unit/helpers/database-mocks');
const { createMockSettingsManager } = require('../unit/helpers/settings-mocks');
const { createMockMaterialPrices } = require('../unit/helpers/test-utils');

const blueprintFixtures = require('../unit/fixtures/blueprints');
const facilitiesFixtures = require('../unit/fixtures/facilities');
const skillsFixtures = require('../unit/fixtures/skills');

describe('Blueprint Calculation Integration Tests', () => {
  let db;

  beforeEach(() => {
    // Create fresh in-memory database for each test
    db = createInMemoryDatabase();
  });

  afterEach(() => {
    if (db) {
      db.close();
    }
  });

  describe('Simple Blueprint Calculation', () => {
    test('calculates materials for Scourge missile blueprint', async () => {
      // Populate database with fixture data
      populateDatabase(db, { blueprint: blueprintFixtures.scourgeBlueprint });

      const facility = facilitiesFixtures.raitaruNoRigs;
      const characterId = null;  // No character-owned blueprints
      const meLevel = 0;
      const runs = 1;

      const result = await calculateBlueprintMaterials(
        blueprintFixtures.TYPE_IDS.SCOURGE_BLUEPRINT,
        runs,
        meLevel,
        characterId,
        facility,
        0,  // depth
        db
      );

      expect(result).toBeDefined();
      expect(result.materials).toBeDefined();
      expect(typeof result.materials).toBe('object');
      expect(Object.keys(result.materials).length).toBeGreaterThan(0);

      // Verify expected materials present (materials is an object with typeID as keys)
      expect(result.materials[34]).toBeDefined();  // Tritanium
      expect(result.materials[34]).toBeGreaterThan(0);
    });

    test('applies ME bonus to material quantities', async () => {
      populateDatabase(db, { blueprint: blueprintFixtures.scourgeBlueprint });

      const facility = facilitiesFixtures.raitaruNoRigs;

      // Calculate with ME 0
      const resultME0 = await calculateBlueprintMaterials(
        blueprintFixtures.TYPE_IDS.SCOURGE_BLUEPRINT,
        1,
        0,
        null,
        facility,
        0,
        db
      );

      // Calculate with ME 10
      const resultME10 = await calculateBlueprintMaterials(
        blueprintFixtures.TYPE_IDS.SCOURGE_BLUEPRINT,
        1,
        10,
        null,
        facility,
        0,
        db
      );

      // ME 10 should reduce material quantities
      if (resultME0.materials && resultME10.materials) {
        const tritME0 = resultME0.materials[34];  // Tritanium
        const tritME10 = resultME10.materials[34];

        if (tritME0 && tritME10) {
          expect(tritME10).toBeLessThanOrEqual(tritME0);
        }
      }
    });

    test('applies facility bonuses', async () => {
      populateDatabase(db, { blueprint: blueprintFixtures.scourgeBlueprint });

      // Calculate with NPC station (no bonuses)
      const resultNPC = await calculateBlueprintMaterials(
        blueprintFixtures.TYPE_IDS.SCOURGE_BLUEPRINT,
        1,
        10,
        null,
        facilitiesFixtures.npcStation,
        0,
        db
      );

      // Calculate with Raitaru (1% ME bonus)
      const resultRaitaru = await calculateBlueprintMaterials(
        blueprintFixtures.TYPE_IDS.SCOURGE_BLUEPRINT,
        1,
        10,
        null,
        facilitiesFixtures.raitaruNoRigs,
        0,
        db
      );

      // Raitaru should have lower or equal material requirements
      if (resultNPC.materials && resultRaitaru.materials) {
        const totalNPC = Object.values(resultNPC.materials).reduce((sum, qty) => sum + qty, 0);
        const totalRaitaru = Object.values(resultRaitaru.materials).reduce((sum, qty) => sum + qty, 0);

        expect(totalRaitaru).toBeLessThanOrEqual(totalNPC);
      }
    });
  });

  describe('Character-Owned Blueprints', () => {
    test('uses character blueprint ME when available', async () => {
      populateDatabase(db, { blueprint: blueprintFixtures.scourgeBlueprint });

      // Mock settings manager with character blueprints
      const mockSettings = createMockSettingsManager({
        owned_blueprints: blueprintFixtures.characterBlueprints,
        characters: [skillsFixtures.basicManufacturingSkills]
      });

      jest.mock('../../src/main/settings-manager', () => mockSettings);

      const facility = facilitiesFixtures.raitaruNoRigs;
      const characterId = 123456;

      const result = await calculateBlueprintMaterials(
        blueprintFixtures.TYPE_IDS.SCOURGE_BLUEPRINT,
        1,
        0,  // ME passed as 0, but character blueprint has ME 10
        characterId,
        facility,
        0,
        db
      );

      // Should use character's ME level
      expect(result).toBeDefined();

      jest.unmock('../../src/main/settings-manager');
    });
  });

  describe('Multiple Runs', () => {
    test('scales material quantities for multiple runs', async () => {
      populateDatabase(db, { blueprint: blueprintFixtures.scourgeBlueprint });

      const facility = facilitiesFixtures.raitaruNoRigs;

      const result1Run = await calculateBlueprintMaterials(
        blueprintFixtures.TYPE_IDS.SCOURGE_BLUEPRINT,
        1,
        0,
        null,
        facility,
        0,
        db
      );

      const result10Runs = await calculateBlueprintMaterials(
        blueprintFixtures.TYPE_IDS.SCOURGE_BLUEPRINT,
        10,
        0,
        null,
        facility,
        0,
        db
      );

      if (result1Run.materials && result10Runs.materials) {
        const trit1 = result1Run.materials[34];  // Tritanium
        const trit10 = result10Runs.materials[34];

        if (trit1 && trit10) {
          // 10 runs should be approximately 10x materials
          // Due to rounding, it may be slightly less (e.g., 49.5 per run × 10 = 495 vs 50 × 10 = 500)
          expect(trit10).toBeGreaterThanOrEqual(trit1 * 9.9);
        }
      }
    });
  });

  describe('Error Handling', () => {
    test('handles invalid blueprint typeID gracefully', async () => {
      const facility = facilitiesFixtures.raitaruNoRigs;

      await expect(
        calculateBlueprintMaterials(
          99999,  // Invalid blueprint
          1,
          0,
          null,
          facility,
          0,
          db
        )
      ).resolves.toBeDefined();
    });

    test('handles null facility', async () => {
      populateDatabase(db, { blueprint: blueprintFixtures.scourgeBlueprint });

      await expect(
        calculateBlueprintMaterials(
          blueprintFixtures.TYPE_IDS.SCOURGE_BLUEPRINT,
          1,
          0,
          null,
          null,  // Null facility
          0,
          db
        )
      ).resolves.toBeDefined();
    });
  });

  describe('Pricing Integration', () => {
    test('calculates total material cost', async () => {
      populateDatabase(db, { blueprint: blueprintFixtures.scourgeBlueprint });

      const facility = facilitiesFixtures.raitaruNoRigs;
      const prices = createMockMaterialPrices();

      const result = await calculateBlueprintMaterials(
        blueprintFixtures.TYPE_IDS.SCOURGE_BLUEPRINT,
        1,
        0,
        null,
        facility,
        0,
        db
      );

      if (result.materials) {
        const totalCost = Object.entries(result.materials).reduce((sum, [typeID, quantity]) => {
          const price = prices[typeID] || 0;
          return sum + (quantity * price);
        }, 0);

        expect(totalCost).toBeGreaterThan(0);
      }
    });
  });

  describe('Snapshot Testing', () => {
    test('material tree structure matches snapshot', async () => {
      populateDatabase(db, { blueprint: blueprintFixtures.scourgeBlueprint });

      const facility = facilitiesFixtures.raitaruNoRigs;

      const result = await calculateBlueprintMaterials(
        blueprintFixtures.TYPE_IDS.SCOURGE_BLUEPRINT,
        1,
        10,
        null,
        facility,
        0,
        db
      );

      // Snapshot test to detect regressions in calculation structure
      expect(result).toMatchSnapshot();
    });
  });
});
