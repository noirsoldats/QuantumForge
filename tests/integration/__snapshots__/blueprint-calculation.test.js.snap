// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Blueprint Calculation Integration Tests Snapshot Testing material tree structure matches snapshot 1`] = `
{
  "breakdown": [
    {
      "blueprintName": "Scourge Light Missile Blueprint",
      "blueprintTypeID": 810,
      "intermediateComponents": [],
      "meLevel": 10,
      "productName": "Scourge Light Missile",
      "productQuantity": 100,
      "productTypeID": 209,
      "rawMaterials": [
        {
          "quantity": 45,
          "typeID": 34,
          "typeName": "Tritanium",
        },
        {
          "quantity": 23,
          "typeID": 35,
          "typeName": "Pyerite",
        },
        {
          "quantity": 5,
          "typeID": 36,
          "typeName": "Mexallon",
        },
        {
          "quantity": 1,
          "typeID": 11399,
          "typeName": "Morphite",
        },
      ],
      "runs": 1,
    },
  ],
  "materials": {
    "11399": 1,
    "34": 45,
    "35": 23,
    "36": 5,
  },
  "pricing": {
    "inputCosts": {
      "allPricesAvailable": true,
      "itemsWithPrices": 4,
      "itemsWithoutPrices": 0,
      "materialPrices": {
        "11399": {
          "hasPrice": true,
          "quantity": 1,
          "totalPrice": 21980,
          "unitPrice": 21980,
        },
        "34": {
          "hasPrice": true,
          "quantity": 45,
          "totalPrice": 450,
          "unitPrice": 10,
        },
        "35": {
          "hasPrice": true,
          "quantity": 23,
          "totalPrice": 756.47,
          "unitPrice": 32.89,
        },
        "36": {
          "hasPrice": true,
          "quantity": 5,
          "totalPrice": 308.8,
          "unitPrice": 61.76,
        },
      },
      "totalCost": 23495.27,
    },
    "jobCost": 997.1618203283948,
    "jobCostBreakdown": {
      "estimatedItemValue": 4597.334349139672,
      "facilityTax": 0,
      "facilityTaxRate": 0,
      "jobBaseCost": 813.2684463628079,
      "jobGrossCost": 813.2684463628079,
      "sccSurcharge": 183.89337396558687,
      "structureCostBonus": 0,
      "systemCostIndex": 0.1769,
      "totalJobCost": 997.1618203283948,
    },
    "outputValue": {
      "hasPrice": true,
      "quantity": 100,
      "totalValue": 5890,
      "typeID": 209,
      "unitPrice": 58.9,
    },
    "profit": -19241.998370328394,
    "profitMargin": -326.6892762364753,
    "salesTax": 639.56655,
    "taxesBreakdown": {
      "accountingSkillLevel": 5,
      "baseSalesTaxRate": 7.5,
      "brokerRelationsSkillLevel": 5,
      "effectiveSalesTaxRate": 3.3749999999999996,
      "materialBrokerFee": 352.42905,
      "materialBrokerFeeRate": 1.5,
      "materialsCost": 23495.27,
      "outputValue": 5890,
      "productBrokerFee": 88.35,
      "productBrokerFeeRate": 1.5,
      "productSalesTax": 198.78749999999997,
      "totalMaterialFees": 352.42905,
      "totalProductFees": 287.13749999999993,
      "totalTaxes": 639.56655,
    },
    "totalCosts": 25131.998370328394,
  },
  "product": {
    "quantity": 100,
    "typeID": 209,
    "typeName": "Scourge Light Missile",
  },
}
`;
